lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart

```diff
@@ -909,14 +964,32 @@ class EmbedderFrameContext {
   ///
   /// These picture recorders will be "claimed" in the paint phase by platform
   /// views being composited into the scene.
-  final List<CkPictureRecorder> pictureRecordersCreatedDuringPreroll =
-      <CkPictureRecorder>[];
+  final Map<PictureLayer, CkPictureRecorder>
+      pictureRecordersCreatedDuringPreroll =
+      <PictureLayer, CkPictureRecorder>{};
 
   /// Picture recorders which were actually used in the paint phase.
   ///
   /// This is a subset of [_pictureRecordersCreatedDuringPreroll].
   final List<CkPictureRecorder> pictureRecorders = <CkPictureRecorder>[];
 
-  /// The number of platform views in this frame.
-  int viewCount = 0;
+  /// The index of the current picture recorder.
+  int pictureRecorderIndex = 0;
+
+  /// List of picture recorders and platform view ids in the order they were
+  /// painted.
+  final List<Object> sceneElements = <Object>[];
```

Comment:
It feels a bit like we might benefit from an intermediate class, possibly polymorphic, that wraps some of the constructs we have here. Something like
```
sealed class SceneElement {
}

class PictureSceneElement implements SceneElement {
    PictureLayer layer;
    CkPictureRecorder recorder;
}

class PlatformViewElement implements SceneElement {
    int viewId;
}
```
And maybe store this in `sceneElements` instead (and maybe it's what the values in the `pictureRecordersCreatedDuringPreroll` should be?). I think this will reduce the amount of mapping/reverse mapping you'll need in the `optimizeRendering` path, and also make the typing less loosey goosey (rather than `List<Object>` you can actually have a more constrained type tree).

Reply:
done